packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol

```diff
@@ -288,6 +288,12 @@ abstract contract CrossDomainMessenger is
         xDomainMsgSender = Constants.DEFAULT_L2_SENDER;
 
         if (success) {
+            // This check is identical to one above, but it ensures that the same message cannot be relayed
+            // twice, and adds a layer of protection against rentrancy.
+            require(
+                successfulMessages[versionedHash] == false,
+                "CrossDomainMessenger: message has already been relayed via reentrancy"
+            );
```

Comment:
A [comment](https://github.com/ethereum-optimism/optimism/pull/8797#discussion_r1442352945) was left on this line in the other PR: 
> I don't see any expect reverts for this case

Reply:
This line should be unreachable. 

Prior to the other fix in this contract (on the initializer), this line would only be reachable during the reinit intercept attack. That makes this change redundant, but IMO it is a good idea as it implements a more FREI-PI like pattern that enforces a pretty basic invariant that no message can be relayed twice, which feels really good to be enforcing. 

However I agree that it's bad practice to have an error message that we can't test. Therefore I will follow the pattern we've sent elsewhere in this contract of using an `assert()` for unreachable conditions:

https://github.com/ethereum-optimism/optimism/blob/jm/fix-l2sender-reinit/packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol#L244-L245